### DISCUSSION TITLE
How to properly set up multipass for high SPP

### DISCUSSION THREAD

**Discussion by Disgruntoad** [mitsuba-renderer/mitsuba3#1284] [Q&A] [answered]
![SPP_OOM](https://github.com/user-attachments/assets/4aa51a69-bce8-47b1-ac36-865ee410a6d7)

Hi all,

I was trying to reduce noise in a custom rendering based on the banner_02 asset (https://mitsuba.readthedocs.io/en/stable/src/gallery.html), and to do so I needed to increase the number of samples per pixel to reduce noise. I ran into the error pictured above (reproduced with the original scene.xml file for reproducibility).

I'm not quite sure what is going wrong. For reference, my system has 64gb ram and 24gb vram. With spp=2048, my system uses under 15gb ram and 2gb vram to render. I assume I'm making a relatively simple mistake, but a google // mitsuba doc search didn't reveal anything. Any help would be appreciated.

I understand that the error is an attempt to allocate 34gb vram, but the render doesn't use more than 2gb vram at any point.

**Comment by merlinND** [comment, marked as answer]
Hello @Disgruntoad,

Have you tried giving a smaller `samples_per_pass` value to the integrator?

There's a warning indicating that it's deprecated:
https://github.com/mitsuba-renderer/mitsuba3/blob/29d65370526d266975237c60fc1586fb509bb048/src/render/integrator.cpp#L109-L115

But I'm not familiar with the new solution, so it's maybe still worth trying.

**Reply by Disgruntoad** [reply]
Edit: Ignore this comment, I misread your response.

Hi @merlinND,

I tried with spp=2048 and it works fine, using less than 15gb system ram and 2gb vram. I'm not sure why spp=4096 causes the large allocation // oom error.

I'm a bit confused as I was following: https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_samplers.html

To clarify, spp is just a parameter setting the `sample_count` in the demo scene.xml file. When I leave that parameter unset, it does generate a render, but it is incredibly noisy and does not converge.

**Reply by Disgruntoad** [reply]
Ah, I realize that I have misread your response. 

I just attempted to set the `samples_per_pass` parameter, but it resulted in the warning you noted & same error message as I got before, except it's not as oom...

`Could not allocate 17179869184 bytes of device memory.` Hmm. Maybe setting it to 1024 is too high.

Yes! If I reduce `samples_per_pass` to 128, it does successfully render. Thank you!

I am getting a new 'Warn' messages, but I believe that is fine.

`2024-08-22 04:34:22 WARN  main  [SamplingIntegrator] render(): forcing 'evaluate=true' since multi-pass rendering was requested.`

Thanks again for the help!

**Reply by merlinND** [reply]
I'm glad it worked! 
I don't think it's the intended solution though, maybe @njroussel or @Speierers know how the automatic batching is supposed to work?

**Reply by njroussel** [reply]
It should "just work :tm:". The idea is that it would figure out how to chop up the total number of samples, given that OptiX has a maxmium number of samples it can handle per kernel.
 
In any case, the memory usage should not increase. I'll have a look.

**Reply by njroussel** [reply]
Had a quick look at it this morning. It's a bit unfortunate but we need to evaluate and store the sampler's state between each pass, just in case it's some stratified sampler. That state can be surprisingly large, for example in an `independent` sampler it's represented by two 8 byte values per lane/thread:
`2 * 8 * (2 ** 31 - 1) ~= 68 GB`
I the case of @Disgruntoad 's example this makes sense: one of the two values would require 34 GB.

We should maybe have a  special code path for the `independent` sampler as it really doesn't need to store its state between passes.

**Reply by Disgruntoad** [reply]
@njroussel 

Thanks for the explanation! I'm a little confused though -- could you break down the math for me? I assume it's
2 values
8 bytes
2**31 - 1 lanes?

What are the lane/threads and how did you arrive at 2**31-1?

Pretty new here, want to start building up some knowledge. Thanks!

**Reply by njroussel** [reply]
Yes, that breakdown is correct. 

The `2**31 - 1` is the kernel width / launch size - it defines the number of threads/elements that needs to be computed. It's a hardware limit, we can't go above that. Of course, in many cases we're well below that value. However, for  high-spp&high-res renderings that limit can be hit quickly and then the image gets rendered in multiple passes (multiple kernels of `2**32-1` each).

**Reply by Disgruntoad** [reply]
Hey, y'all. I just wanted to revisit this. I recently got this error message:
![Screenshot from 2025-06-06 02-44-32](https://github.com/user-attachments/assets/8ff5f32d-1c22-476d-9c74-169e49d94232)

I just wanted to confirm that by "render using multiple passes" it means, effectively

```
num_passes = 10
image = 0
for i in range(num_passes):
    image += mi.render(scene, sampler_seed=i)
image /= num_passes
```

It seems to work well in practice, but am I missing anything or is there a built-in way to do this?

**Reply by merlinND** [reply]
Hello @Disgruntoad,

Yes, your example looks right except that you might want to:
1. Call `dr.eval(image)` in the loop
2. Possibly multiply the rendered image by `1/num_passes` in the loop directly instead of at the end, depending on numerics.
